src/logic/addressBook/utils/v2-migration.ts

```diff
@@ -0,0 +1,130 @@
+import { AddressBookEntry, AddressBookState, makeAddressBookEntry } from 'src/logic/addressBook/model/addressBook'
+import { saveSafes, StoredSafes } from 'src/logic/safe/utils'
+import { getNetworkName } from 'src/config'
+import { getWeb3 } from 'src/logic/wallets/getWeb3'
+import { Errors, logError } from 'src/logic/exceptions/CodedException'
+import { getEntryIndex, isValidAddressBookName } from '.'
+
+interface StorageConfig {
+  states: string[]
+  namespace: string
+  namespaceSeparator: string
+}
+
+/**
+ * Migrates the safes names from the Safe Object to the Address Book
+ *
+ * Works on the persistence layer, reads from the localStorage and stores back the mutated safe info through immortalDB.
+ *
+ * @note If the Safe name is an invalid AB name, it's renamed to "Migrated from: {safe.name}"
+ */
+const migrateSafeNames = ({ states, namespace, namespaceSeparator }: StorageConfig): void => {
+  const prefix = `v2_${getNetworkName()}`
+  const safesKey = `_immortal|${prefix}__SAFES`
+  const storedSafes = localStorage.getItem(safesKey)
+
+  if (!storedSafes) {
+    // nothing left to migrate
+    return
+  }
+
+  const parsedStoredSafes = JSON.parse(storedSafes) as Record<string, any>
+
+  if (Object.entries(parsedStoredSafes).every(([, { name }]) => name === undefined)) {
+    // no name key, safes already migrated
+    return
+  }
+
+  // make address book entries from the safe names & addresses
+  const safeAbEntries: AddressBookState = Object.values(parsedStoredSafes)
+    .filter(({ name }) => name && isValidAddressBookName(name))
+    .map(({ address, name }) => makeAddressBookEntry({ address, name }))
+
+  // remove names from the safes in place
+  Object.values(parsedStoredSafes).forEach((item) => {
+    item.owners = item.owners.map((owner: any) => owner.address)
+    delete item.name
+  })
+  const migratedSafes = parsedStoredSafes as StoredSafes
+
+  const [state] = states
+  const addressBookKey = `${namespace}${namespaceSeparator}${state}`
+  const storedAddressBook = localStorage.getItem(addressBookKey)
+  const addressBookToStore: AddressBookState = storedAddressBook ? JSON.parse(storedAddressBook) : []
+
+  // mutate `addressBookToStore` by adding safes' information
+  safeAbEntries.forEach((entry) => {
+    const safeIndex = getEntryIndex(addressBookToStore, entry)
+
+    if (safeIndex >= 0) {
+      // update AB entry with what was stored in the safe object
+      addressBookToStore[safeIndex] = entry
+    } else {
+      // add the safe entry to the AB
+      addressBookToStore.push(entry)
+    }
+  })
+
+  // store the mutated address book
+  localStorage.setItem(addressBookKey, JSON.stringify(addressBookToStore))
+
+  // update stored safe
+  localStorage.setItem(safesKey, JSON.stringify(migratedSafes))
```

Comment:
@fernandomg as you see, I've added a redundant call to set the storage item directly, so that we don't run in any async issues with the `saveSafes` (on the following line).

Reply:
smart!

Reply:
Is there a possibility for some race condition with immortaldb?

Reply:
@mikheevm this migration is executed on store import time, and ImmortalDB kicks in after that (in the main `index.ts`).

ImmortalDB's self-healing mechanism is quite... interesting, and we want to be 100% on the safe side, I would disable the IndexedDB storage. And in future ultimately replace ImmortalDB with just localStorage for simplicity.

Reply:
@mikheevm do you mean immortalDB overwriting localStorage with the old data that's in indexedDB?

Reply:
I'll merge this PR now to unblock further development on the epic branch. Let's discuss this what we will do with ImmortalDB in #2346.